sdk/core/Azure.Core/src/Pipeline/RetryPolicy.cs

```diff
@@ -0,0 +1,331 @@
+        }
+
+        /// <summary>
+        /// This method can be overriden to control whether a request should be retried. It will be called for any response where
+        /// <see cref="Response.IsError"/> is true, or if an exception is thrown from any subsequent pipeline policies or the transport.
+        /// This method will only be called for sync methods.
+        /// </summary>
+        /// <param name="message">The message containing the request and response.</param>
+        /// <returns>Whether or not to retry.</returns>
+        protected virtual bool ShouldRetry(HttpMessage message) => ShouldRetryInternal(message);
+
+        /// <summary>
+        /// This method can be overriden to control whether a request should be retried.  It will be called for any response where
+        /// <see cref="Response.IsError"/> is true, or if an exception is thrown from any subsequent pipeline policies or the transport.
+        /// This method will only be called for async methods.
+        /// </summary>
+        /// <param name="message">The message containing the request and response.</param>
+        /// <returns>Whether or not to retry.</returns>
+        protected virtual ValueTask<bool> ShouldRetryAsync(HttpMessage message) => new(ShouldRetryInternal(message));
+
+        private bool ShouldRetryInternal(HttpMessage message)
+        {
+            if (message.RetryNumber < _maxRetries)
+            {
+                if (message.LastException != null)
+                {
+                    return message.ResponseClassifier.IsRetriable(message, message.LastException);
+                }
+
+                // Response.IsError is true if we get here
+                return message.ResponseClassifier.IsRetriableResponse(message);
+            }
+
+            // out of retries
+            return false;
+        }
+
+        /// <summary>
+        /// This method can be overriden to control how long to delay before retrying. This method will only be called for sync methods.
+        /// </summary>
+        /// <param name="message">The message containing the request and response.</param>
+        /// <returns>The amount of time to delay before retrying.</returns>
+        protected virtual TimeSpan CalculateNextDelay(HttpMessage message) => CalculateNextDelayInternal(message);
+
+        /// <summary>
+        /// This method can be overriden to control how long to delay before retrying. This method will only be called for async methods.
+        /// </summary>
+        /// <param name="message">The message containing the request and response.</param>
+        /// <returns>The amount of time to delay before retrying.</returns>
+        protected virtual ValueTask<TimeSpan> CalculateNextDelayAsync(HttpMessage message) => new(CalculateNextDelayInternal(message));
+
+        /// <summary>
+        /// This method can be overriden to introduce logic before the request is sent. This will run even for the first attempt.
```

Comment:
```suggestion
        /// This method can be overridden to introduce logic before each request attempt is sent. This will run even for the first attempt.
```